source/hash_table.c

```diff
@@ -653,38 +671,60 @@ static inline void s_get_next_element(struct aws_hash_iter *iter, size_t start_s
         if (entry->hash_code) {
             iter->element = entry->element;
             iter->slot = i;
+            iter->status = AWS_HASH_ITER_STATUS_READY_FOR_USE;
```

Comment:
Is this a bug fix? Please mention in the commit message if so

Reply:
This is more of an improvement than a bug fix.  There is an implicit state machine for using an ```aws_hash_iter```: for example, it is an error to call ```iter_delete``` on an iter which is in the ```done``` state, and after ```iter_delete```  has been called, the only valid thing to do is to call ```iter_next```.  By making the state machine explicit, it is possible to both test it during debug builds, and to prove properties about consumers of the hash-table.  In particular, this sets up aws-encryption-sdk proofs to check this property.

About the commit message: all commits will be squashed when the PR is merged.  I will write a better commit message for the squash commit.